### ISSUE TITLE
latex support

### ISSUE THREAD

**Issue by Cyberfly100** [plotly/dash#242]
I tried to use latex in dash, but it is not working. 
It seems that the mathjax javacript library is not loaded.

**Comment by chriddyp** [comment]
I wonder if the solution is just to add MathJax to https://github.com/plotly/dash-core-components/blob/d034e25f6f56423ef2de7e5905328396040cf4d8/dash_core_components/__init__.py#L25-L29

**Comment by jessexknight** [comment]
It is possible to render MathJax in static Dash content, but not in dynamic content.

Here is a MWE

```
import dash
import dash_html_components as html

app = dash.Dash(__name__)
mathjax = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-MML-AM_CHTML'
app.scripts.append_script({ 'external_url' : mathjax })

app.layout = html.Div(id='main',children=[
  html.Div(id='static',children='$$ x=1 $$'),
  html.Div(id='dynamic',children=''),
  html.Button('Add Math',id='button'),
])

@app.callback(
  dash.dependencies.Output('dynamic','children'),
 [dash.dependencies.Input('button','n_clicks')]
)
def addmath(n_clicks):
  if n_clicks:
    return '$$ x=1 $$'
  else:
    return None

if __name__ == '__main__':
  app.run_server(debug=True)
```

[related community.plot.ly thread](https://community.plot.ly/t/mathjax-in-dash/)

**Comment by aztan2** [comment]
Hi, I'm wondering if there has been progress on this issue? Specifically, I would like to be able to render latex in the axis titles/labels of a graph.

**Comment by jessexknight** [comment]
There is a [pull request](https://github.com/plotly/dash-core-components/pull/194) in progress to solve this issue, but it seems like it has stalled at the moment.

**Comment by xhlulu** [comment]
I created dash-katex, a library that let you render latex inside a component `DashKatex`, obviously using katex js. This does not solve the problem of rendering latex inside a plotly graph object, but lets you dynamically run latex inside divs.

Link: https://github.com/xhlulu/dash-katex

**Comment by vdkotian** [comment]
Is this issue still available for development? I am seeking to contribute.

**Comment by renatobellotti** [comment]
Is there any progress on this? Latex support is quite important for scientific dashboards...

**Comment by renatobellotti** [comment]
What about now?

**Comment by yueyericardo** [comment]
After searching around, I found [Latex to Image API](https://math.now.sh/home) very useful.
And I wrote 2 function to make it work with `$a+b$` and `$$a+b$$` for dcc.markdown.

```python
import dash_core_components as dcc
import dash_html_components as html
import dash
import urllib.parse
import re


def toimage(x):
    if x[1] and x[-2] == r'$':
        x = x[2:-2]
        return "\n<img src='https://math.now.sh?from={}' style='display: block; margin: 0.5em auto;'>\n".format(urllib.parse.quote_plus(x))
    else:
        x = x[1:-1]
        return r'![](https://math.now.sh?from={})'.format(urllib.parse.quote_plus(x))


def convert(text):
    return re.sub(r'\${2}([^$]+)\${2}|\$(.+?)\$', lambda x: toimage(x.group()), text)


app = dash.Dash()

Markdown_text = r"""
Let's see if it works:  
$$\hat P \psi_k(x) =p \psi_k(x)$$ 

$$-i\hbar \frac{\partial {c\ e^{ikx}}}{\partial x} =-i\hbar\ c\ ik\ e^{ikx} $$ 

$$\hbar k\ c\ e^{ikx} = \hbar k\ \psi_k(x) \tag{2}$$
with $p=\hbar k$
"""

Markdown_text = convert(Markdown_text)


app.layout = html.Div([
    dcc.Markdown(Markdown_text, dangerously_allow_html=True)
])

if __name__ == '__main__':
    app.run_server(host='0.0.0.0')
```
![](https://yyrcd-1256568788.cos.na-siliconvalley.myqcloud.com/yyrcd/2020-03-20-214750.png)

**Comment by renatobellotti** [comment]
Does this also work for axis labels?

**Comment by jessexknight** [comment]
I will look into this in the next 2 weeks. For now, here are some ideas somebody may want to try:

- load mathjax as suggested [above](https://github.com/plotly/dash/issues/242#issuecomment-391708460)
- define and include javascript function to re-render some element by mathjax: [examples](https://github.com/mathjax/MathJax-demos-web#processing-user-input)
- add a [clientside callback](https://dash.plotly.com/clientside-callbacks) in plotly to trigger the js function to re-render

**Comment by yueyericardo** [comment]
> Does this also work for axis labels?

Hey, I just found some workaround for axis labels, as well as latex in markdown.
I'll put this into a repo soon and share how to use this.

An example
![image](https://user-images.githubusercontent.com/9999318/79058636-6440b400-7c3e-11ea-9e9a-0e4f48d0b66e.png)


Richard

**Comment by renatobellotti** [comment]
> > Does this also work for axis labels?
> 
> Hey, I just found some (tricky) workaround for axis labels, as well as latex in markdown.
> I'll put this into a repo soon and share how to use this.
> 
> An example
> ![image](https://user-images.githubusercontent.com/9999318/79058636-6440b400-7c3e-11ea-9e9a-0e4f48d0b66e.png)
> 
> Richard

This looks great! How did you do it? And is this something you could provide a pull request for? ;)

**Comment by yueyericardo** [comment]
> This looks great! How did you do it? And is this something you could provide a pull request for? ;)

Hey sorry for the delay, I just created a complete example  https://github.com/yueyericardo/dash_latex explain how to make it work.

Example2 will come later.

**Comment by jessexknight** [comment]
Well, now I'm confused. It looks like dynamic MathJax in axes labels now works. I thought that this was not working before. Can anybody confirm that this was part of the problem and now fixed?

This MWE will compile the user-input LaTeX live and render the result on y-axis label.
```python
import dash
import dash_html_components as html
import dash_core_components as dcc
from dash.dependencies import Input, Output

mathjax = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-MML-AM_CHTML'
app = dash.Dash(__name__,
  external_scripts=[mathjax],
)

def fig_fun(label='default'):
  return {
    'data': [{'x': [-5, 0,+5], 'y': [0,0,5] }],
    'layout': {'yaxis': {'title': label}},
  }

ylabel = '$y = \\begin{cases} 0 & x < 0 \\\\ x & x \ge 0 \\end{cases}$'

app.layout = html.Div(id='main',children=[
  dcc.Input(id='input',value=ylabel,style={'width':'100%'}),
  dcc.Graph(id='graph',figure=fig_fun())
])

@app.callback( Output('graph','figure'), [Input('input','value')] )
def update_fun(label):
  return fig_fun(label=label)

if __name__ == '__main__':
  app.run_server(debug=True)
```

![dash-axis-live](https://user-images.githubusercontent.com/14393754/79508302-4a6ceb80-8007-11ea-8104-8ca8ff195372.jpg)

**Comment by jessexknight** [comment]
Okay. Here is another hack that will probably solve **all** dynamic LaTeX rendering issues but you have to decide the refresh rate (trading speed vs CPU).

After including MathJax like above:
```python
app = dash.Dash(__name__, external_scripts=[
  'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-MML-AM_CHTML',
])
```
add the file `./assets/mathjax-refresh.js` with the following content:
```js
setInterval("MathJax.Hub.Queue(['Typeset',MathJax.Hub])",1000);
```
This asks MathJax to scan the page for any new `$...$` every 1000 milliseconds, and re-render it. It solved all my MWE here without any extra modification of the Python code. You may want to increase the refresh rate by decreasing `1000`. I tried `1` and it was very fast, but CPU 10% `¯\_(ツ)_/¯`.

more info:

 - [about Dash assets](https://dash.plotly.com/external-resources)
 - [about MathJax.Hub.Queue](https://docs.mathjax.org/en/v1.0/typeset.html)
 - [about setInterval](https://www.w3schools.com/js/js_timing.asp)

**Comment by renatobellotti** [comment]
That workaround seems to work, thank you so much!

Just one thing is a bit suboptimal, but that's a luxury problem. When I set the axis label to something like `$\sigma_x$ [mm]`, only the sigma_x part is rendered, and the units are omitted.

Workaround:

Use  `$\sigma_x \mathrm{[mm]}$` instead.

**Comment by jessexknight** [comment]
Ugh. It seems the `mathjax-refresh` hack somehow interferes with the dynamic axis labelling, making it very slow. Apologies that I didn't test this enough before.

@renatobellotti I wouldn't call that a luxury problem. Mixing TeX and non-TeX might be quite common, and we shouldn't have to wrap all non-TeX in `\textrm{...}`.

Sometime in the past year it seems TeX support was added to axis labels, but I can't find any commit that references this. Essentially I'd like to turn that off, since I'd prefer the refresh hack.

**Comment by Davide-sd** [comment]
@jessexknight , thank you for your solution, it works nice if there is a Graph output! 

However, when there is no Graph output (but the app has some other outputs), then it seems MathJax is not loaded and the latex code in other controls is not rendered. Is there a quick workaround?

**Comment by jessexknight** [comment]
I don't have time to look into this right now but [this](https://github.com/plotly/plotly.js/issues/4563) may be useful.

**Comment by nicolaskruchten** [comment]
To get starndard Plotly.s LaTeX support in figure titles and axis labels, it looks like doing the Dash equivalent of [what's in the Plotly.js readme](https://github.com/plotly/plotly.js/blob/master/dist/README.md) does work:

1. add `external_scripts=["https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-MML-AM_CHTML" ]` to the `app = dash.Dash()` call
2. add a JS file to `assets` containing `window.PlotlyConfig = {MathJaxConfig: 'local'};`

**Comment by nicolaskruchten** [comment]
> When I set the axis label to something like $\sigma_x$ [mm], only the sigma_x part is rendered, and the units are omitted.

This is sort of a known feature of how Plotly.js works at the moment: labels in figures are either all-LaTex or all-not-LaTeX.

**Comment by haduong83** [comment]
Hi,

I tried to add the configuration into my app:

mathJaxSxrip = ["https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML" ]
# external JavaScript files

server = Flask(__name__, instance_relative_config=False)
# server = app.server
app = dash.Dash(server=server, external_scripts=mathJaxSxrip, external_stylesheets = [dbc.themes.BOOTSTRAP])
app.config.suppress_callback_exceptions = True
app.scripts.config.serve_locally = False

and render the following equation:

html.Div(children="Render formular \(\lim_{t \\rightarrow \infty} \pi = 0\)
$$\lim_{t \\rightarrow \infty} \pi = 0$$")

but it is not working. I am using the following dash version:

dash==1.19.0
dash-bootstrap-components==0.11.3
dash-core-components==1.15.0
dash-html-components==1.1.2
dash-renderer==1.9.0

Any ideas on what could be wrong?

Thanks,

**Comment by wvwhome** [comment]
MathJax and Plotly have mystery.  I got this to work by adding an 'assets' folder as further posted by jessexknight.  The bigger question is whether/when Plotly will provide official support for Mathjax in general.

**Comment by nopria** [comment]
Have you looked into this?
https://chrisvoncsefalvay.com/2020/07/25/dash-latex/

**Comment by chrisvoncsefalvay** [comment]
Hey @haduong83, I think you're not invoking the script properly. Try checking [here](https://chrisvoncsefalvay.com/2020/07/25/dash-latex/) and take a look at the heading "Integrating MathJax into the template".

**Comment by archmoj** [comment]
Resolved by [v2.3.0](https://github.com/plotly/dash/releases/tag/v2.3.0).
